---
title: "Cattle Paper"
author: "Anderson Freitas"
output: html_notebook
---

```{r}
#Libraries

library(dada2)
library(phyloseq)
library(tidyverse)
library(microbiome)
library(hrbrthemes)
library(gcookbook)
library(dplyr)
library(jeevanuDB)
library(ggplot2)
library(ggpubr)
library(ALDEx2)
library(vegan)
library(knitr)
library(hues)
library(kableExtra)
library(data.table)
library(ggvenn)
library(DataCombine)
library(broom)
library(dunn.test)
library(microeco)
library(magrittr)
library(devtools)
library(SpiecEasi)
library(grid)
library(cowplot)

```

```{r}
#To make thing pretty

theme_set(theme_bw())
pal = "Set1"
scale_colour_discrete <-  function(palname=pal, ...){
  scale_colour_brewer(palette=palname, ...)
}
scale_fill_discrete <-  function(palname=pal, ...){
  scale_fill_brewer(palette=palname, ...)
}

ze=c("#9838d8",     "#9ee544",     "#5f52d7",     "#ddd03e",     "#40207c",     "#6ae184",    "#d84ccd",     "#67af35",     "#8c3ba1",     "#ccdc88",     "#d44290",     "#6fe2c1",     "#df462a",   "#76cadf",     "#d5425b",     "#579c56",     "#c082dc",     "#8f8e33",     "#5767b9",     "#d97f35", "#7f9add",     "#d2a34d",     "#422b61",     "#c6dac0",     "#371d2f",     "#60a78e",     "#88315c",   "#386123",     "#d782b0",     "#2b3722",     "#cfb7d3",     "#9b4528",     "#5e89a4",     "#622b24",   "#bda685",     "#354a62",     "#d5867a",     "#537260",     "#8c677c",     "#705d2c")

ze2 = c("#9cd848",
"#60b7ff",
"#9c261f")

```


# Creating a phyloseq object:

```{r}

# Importing files

seqtab = readRDS("E:/Anderson-BackUp/Seqs/All/seqtab_final.rds")
taxa = readRDS("E:/Anderson-BackUp/Seqs/All/taxonomy_genera.rds")
map <- "E:/Anderson-BackUp/Seqs/All/maptudo.txt"

# Building a phyloseq object

ps <- phyloseq(otu_table(seqtab, taxa_are_rows=FALSE), tax_table(taxa))

sample_metadata = import_qiime_sample_data(map)
input =merge_phyloseq(ps, sample_metadata)
input

```


```{r}

summarize_phyloseq(input)

```

```{r}

#Reads distribution

sdt = data.table(as(sample_data(input), "data.frame"),
                 TotalReads = sample_sums(input), keep.rownames = TRUE)

#Coverage calculation

Good <-function (a) {
  freq.one <-colSums((a)==1)
  freq.one
  num.seq<-colSums(a)
  num.seq.mat<-as.matrix(num.seq)
  nseq<-num.seq.mat
  good.res<-(1-(freq.one/nseq))
  good.res
  return(good.res)
}

Coverage=Good(t(otu_table(ps)))

tab1= cbind(sdt,Coverage)
names(tab1)[names(tab1) == "V1"] <- "Coverage"

#Showing the results
kable(tab1[ ,c(2:5, 15:18)], caption = "Distribution of reads per sample")

```


```{r}
#View(sdt)
#S7017 has low amount. Taking it off.

input2 = subset_samples(input, SampleID != "S7017")
#summarize_phyloseq(input2)

set.seed(2125)
inputR = rarefy_even_depth(input2, sample.size = 1028, replace = FALSE)

#transform to compositional
#clr tranformation -- comes with a pseudocount to avoid zeros.
bac_comp <- microbiome::transform(inputR, "clr")
bac_comp

```

```{r}
#Tests for variables#

#Normality

library(readr)
map_t <- read_csv("E:/Anderson-BackUp/Seqs/All/maptudo2.csv")

shp.adg <- shapiro.test(map_t$ADG)
shp.omd <- shapiro.test(map_t$Average_OMD)
shp.cpi <- shapiro.test(map_t$Average_CPI)

shp.adg
shp.omd
shp.cpi

#If p-value > or = to 0.05 the results DO NOT FOLLOW the normal distribution

#So we must use Kruskal-Wallis test

kw.adg = kruskal.test(ADG ~ SampleType, data = map_t)
kw.omd = kruskal.test(Average_OMD ~ SampleType, data = map_t)
kw.cpi = kruskal.test(Average_CPI ~ SampleType, data = map_t)

kw.adg
kw.omd
kw.cpi

```

```{r}

# Calculating Alpha Diversity

observed=microbiome::alpha(inputR, index = "observed")
meta=microbiome::meta(inputR)
alpha= cbind(observed,meta)

#verify normality
shp <- shapiro.test(alpha$observed)
shp
p_value <- signif(shp$p.value, digits = 3)
p <- if(p_value > 0.05) {"follow"} else {"do not follow"}
p

#Testing difference

attach(alpha) 
kruskal.test(alpha$observed ~ alpha$SampleType, data = alpha)

dunn.test::dunn.test(alpha$observed, g=alpha$SampleType, method="bonferroni", kw=TRUE, label=TRUE, 
      wrap=FALSE, table=F, list=T, rmc=FALSE, alpha=0.05, altp=FALSE)

comparison = list(c("Feces", "Ruminal_Fluid"), c("Saliva", "Ruminal_Fluid"), c("Saliva", "Feces"))

alpha_means = c("Observed")
p= plot_richness(inputR, x= "SampleType", measures= alpha_means, color = "SampleType", )
p + geom_boxplot() +
  labs(x = "Sample Type", y= "Observed Values") +
  geom_jitter(shape=16, position=position_jitter(0)) +
  stat_compare_means(comparisons = comparison) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5)) +
  labs(color='Sample Type') + guides(color = "none")

#setwd("C:/Users/Anderson Freitas/Desktop/Seqs/Figures")
#dev.print(tiff, "fig1_alpha_diversity.tiff", compression = "lzw", res=600, height=13, width=10, units="cm")

```

```{r}
#Permanova

df = as(sample_data(bac_comp), "data.frame")
ds = phyloseq::distance(bac_comp, method = "euclidean")
permanova = adonis(ds ~ SampleType, data = df, permutations = 999)

kable(permanova$aov.tab[ ,1:6], caption = "PERMANOVA")

# Ploting beta diversity graph
input_ord = ordinate(bac_comp, "PCoA" , "euclidean") 
p1 = plot_ordination(bac_comp, input_ord, color = "SampleType")
p1 + geom_point(size = 6, alpha = 0.5) +
  labs(subtitle = "R-squared = 0.10; p = 0.001", color='Sample Type')

#dev.print(tiff, "2_beta_diversity.tiff", compression = "lzw", res=600, height=12, width=16.5, units="cm")

```

```{r}
#Venn diagram 

#Genus

saliva = subset_samples(inputR, SampleType=="Saliva")
rumen = subset_samples(inputR, SampleType=="Ruminal_Fluid")
feces = subset_samples(inputR, SampleType=="Feces")

saliva.gen = aggregate_rare(saliva, level = "Genus", detection = 1/100, prevalence = 1/100)
rumen.gen = aggregate_rare(rumen, level = "Genus", detection = 1/100, prevalence = 1/100)
feces.gen = aggregate_rare(feces, level = "Genus", detection = 1/100, prevalence = 1/100)

saliva.gen <- as.data.frame(tax_table(saliva.gen))
rumen.gen <- as.data.frame(tax_table(rumen.gen))
feces.gen <- as.data.frame(tax_table(feces.gen))

saliva.gen = as.character(saliva.gen$Genus)
rumen.gen = as.character(rumen.gen$Genus)
feces.gen = as.character(feces.gen$Genus)

x <- list(
  saliva.gen = sample(saliva.gen), 
  rumen = sample(rumen.gen), 
  feces.gen = sample(feces.gen)
)

genus = ggvenn(x, fill_color = ze2, stroke_size = 0.5, set_name_size = 5, text_size = 2.5)

d = ggpubr::ggarrange(genus, ncol=1, nrow = 1, align = "hv", common.legend = TRUE)

genus.level = ggpubr::annotate_figure(d,
                                       top = text_grob("Genus level", color = "black", face = "bold", size = 16))

#phylum

saliva = subset_samples(inputR, SampleType=="Saliva")
rumen = subset_samples(inputR, SampleType=="Ruminal_Fluid")
feces = subset_samples(inputR, SampleType=="Feces")

saliva.phy = aggregate_rare(saliva, level = "Phylum", detection = 1/100, prevalence = 1/100)
rumen.phy = aggregate_rare(rumen, level = "Phylum", detection = 1/100, prevalence = 1/100)
feces.phy = aggregate_rare(feces, level = "Phylum", detection = 1/100, prevalence = 1/100)

saliva.phy <- as.data.frame(tax_table(saliva.phy))
rumen.phy <- as.data.frame(tax_table(rumen.phy))
feces.phy <- as.data.frame(tax_table(feces.phy))

saliva.phy = as.character(saliva.phy$Phylum)
rumen.phy = as.character(rumen.phy$Phylum)
feces.phy = as.character(feces.phy$Phylum)

x <- list(
  saliva.phy = sample(saliva.phy), 
  rumen.phy = sample(rumen.phy), 
  feces.phy = sample(feces.phy)
)

phylum = ggvenn(x, fill_color = ze2, stroke_size = 0.5, set_name_size = 5, text_size = 2.5)
phylum

c = ggpubr::ggarrange(phylum, ncol=1, nrow = 1, align = "hv", common.legend = TRUE)

phylum.level = ggpubr::annotate_figure(c,
                                 top = text_grob("Phylum level", color = "black", face = "bold", size = 16))

tudo = ggpubr::ggarrange(phylum.level, genus.level, ncol=2, nrow = 1, align = "hv", common.legend = TRUE)

tudo

#dev.print(tiff, "3_venn_diagram.tiff", compression = "lzw", res=600, height=14, width=21, units="cm")

```

```{r}
#top 20 phylum distribution

ph.fun = tax_glom(inputR, taxrank = "Phylum")

phy20.clr <- microbiome::transform(ph.fun, "clr")

phy20 = names(sort(taxa_sums(phy20.clr), TRUE)[1:21])
top20 = prune_taxa(phy20, phy20.clr)

map=as(sample_data(top20),"data.frame")
otu=t(otu_table(top20))
taxs=tax_table(top20) %>% as("matrix") %>% data.frame()
#name the OTUs according to the toxonomy (keeping individual names)
rownames(otu)=taxs$Phylum
#put toghether counts and metadata
otu.map=data.frame(t(otu),map)

#plotting
myplots=list()
top20=sort(as.factor(taxs$Phylum))
top20=as.matrix(top20)
top20[6,] = "Deinococcus.Thermus"  #that's a gambiarra, do not use it shit on your data
top20=as.data.frame(top20)

# Create replacements data frame
Replaces <- data.frame(from = c("p__"), to = c("")) #not pretty sure about what it does, but works
Replaces

# Replace patterns and return full data frame
top20.l <- FindReplace(data = top20, Var = "V1", replaceData = Replaces,
                       from = "from", to = "to", exact = FALSE)
top20 <- sort(as.factor(top20.l$V1))

for (i in top20) {
  p1=ggboxplot(otu.map, x="SampleType", y=as.character(i), color = "SampleType", 
               palette = "ze2", title = as.character(i))+ xlim("Saliva", "Ruminal_Fluid", "Feces")+ 
    xlab("")+ theme(axis.text.x = element_blank(),
                    axis.text.y = element_text(size = 9),
                    title = element_text(size = 9, hjust = .2, face = "italic"),
                    legend.title = element_blank(),
                    legend.position="bottom")+
    ylab("")
  myplots[[i]]<-p1
}

myplots$Cyanobacteria = NULL


ggarrange(plotlist = myplots, common.legend = T) %>% 
  annotate_figure(left = text_grob("Centered log ratio abundance",
                                   rot = 90))

bbb = ggarrange(plotlist = myplots, common.legend = T, legend = "bottom") %>% 
  annotate_figure(left = text_grob("clr Transformed Abundance",
                                   rot = 90), bottom = text_grob("Sample Type", rot = 0))
#bbb
#dev.print(tiff, "4_box_phyla.tiff", compression = "lzw", res=600, height=8, width=10, units="in")

```

```{r}

df2 <- otu.map %>% gather(key, value, -SampleType) %>% 
  group_by(key) %>% 
  do(tidy(kruskal.test(x= .$value, g = .$SampleType)))

df2

kable(df2, caption = "Are the mean CLR value for phyla different among sample types?", format = "markdown")

#And the Dunn test post hoc

for (i in 1:21) {
  print(colnames(otu.map[i]))
  dunn.test(x=otu.map[,i], g=otu.map$SampleType,
                       method="bonferroni", kw=TRUE, label=TRUE,
                       wrap=FALSE, table=F, list=T, rmc=FALSE, alpha=0.05, altp=FALSE)
}

```

```{r}
## Networks

# Saliva

set.seed(2125)

saliva.net = subset_samples(inputR, SampleType == "Saliva")

sample_info_16S     = sample_data(saliva.net)  %>% as("matrix") %>% data.frame()
otu_table_16S       = t(otu_table(saliva.net)) %>% as("matrix") %>% data.frame()
taxonomy_table_16S  = tax_table(saliva.net)    %>% as("matrix") %>% data.frame()


```

```{r}
# Rumen

set.seed(123)

rumen.net = subset_samples(inputR, SampleType == "Ruminal_Fluid")

rumen.net2 = aggregate_rare(rumen.net, level = "Genus", detection = 1/100, prevalence = 1/100)

sample_info     = sample_data(rumen.net2) %>% as("matrix") %>% data.frame()
otu_table       = otu_table(rumen.net2) %>% as("matrix") %>% data.frame()
taxonomy_table  = tax_table(rumen.net2) %>% as("matrix") %>% data.frame()

```


```{r}

# Feces

set.seed(123)

feces.net = subset_samples(inputR, SampleType == "Feces")

sample_info_16S     = sample_data(feces.net) %>% as("matrix") %>% data.frame()
otu_table_16S       = t(otu_table(feces.net)) %>% as("matrix") %>% data.frame()
taxonomy_table_16S  = tax_table(feces.net) %>% as("matrix") %>% data.frame()


```


```{r}
### Tables

saliva = aggregate_rare(saliva, level = "Genus", detection = 1/100, prevalence = 1/100)
rumen = aggregate_rare(rumen, level = "Genus", detection = 1/100, prevalence = 1/100)
feces = aggregate_rare(feces, level = "Genus", detection = 1/100, prevalence = 1/100)

## Saliva
set.seed(4040)
saliva.gmd = subset_samples(saliva, Comparison == "Yes")

GMD_sal = prune_taxa(taxa_sums(saliva.gmd) > 2, saliva.gmd)
mi1=(as.data.frame((otu_table(GMD_sal))))
var = sample_data(GMD_sal)
treat = as.character(var$ADG_Class)

x <- aldex(mi1, treat, mc.samples=128, test="t", effect=TRUE,
           include.sample.summary=TRUE, denom="iqlr", verbose=TRUE)

aldex=cbind(x, tax_table(GMD_sal))
res2=(aldex[(aldex$we.ep<="0.1"),])
Genus = res2$Genus
res2 = cbind(Genus, res2)
rownames(res2) = NULL
c1 = res2[,1]
c2 = res2[,3:4]
c3 = res2[,25:27]
res3 = cbind(c1,c2,c3)
colnames(res3)
colnames(res3) = c("Genus", "rab.High", "rab.Low", "effect", "overlap", "we.ep") 

round_df <- function(x, digits) {
  # round all numeric variables
  # x: data frame 
  # digits: number of digits to round
  numeric_columns <- sapply(x, mode) == 'numeric'
  x[numeric_columns] <-  round(x[numeric_columns], digits)
  x
}

res4 = round_df(res3, 2)
res4

## Rumen

rumen.gmd = subset_samples(rumen, Comparison == "Yes")

GMD_rum = prune_taxa(taxa_sums(rumen.gmd) > 2, rumen.gmd)
mi2=(as.data.frame(otu_table(GMD_rum)))
var2 = sample_data(GMD_rum)
treat2 = as.character(var2$ADG_Class)

x2 <- aldex(mi2, treat2, mc.samples=128, test="t", effect=TRUE,
            include.sample.summary=TRUE, denom="iqlr", verbose=TRUE)

aldex.rum=cbind(x2, tax_table(GMD_rum))

res.rum=(aldex.rum[(aldex.rum$effect>="0.5"),])

Genus = res.rum$Genus
res.rum = cbind(Genus, res.rum)
rownames(res.rum) = NULL
c1 = res.rum[,1]
c2 = res.rum[,3:4]
c3 = res.rum[,17:19]
res3 = cbind(c1,c2,c3)
colnames(res3)
colnames(res3) = c("Genus", "rab.High", "rab.Low", "effect", "overlap", "we.ep") 

round_df <- function(x, digits) {
  # round all numeric variables
  # x: data frame 
  # digits: number of digits to round
  numeric_columns <- sapply(x, mode) == 'numeric'
  x[numeric_columns] <-  round(x[numeric_columns], digits)
  x
}


res4.rum = round_df(res3, 2)
res4.rum

## Feces

feces.gmd = subset_samples(feces, Comparison == "Yes")

GMD_fec = prune_taxa(taxa_sums(feces.gmd) > 2, feces.gmd)
mi3=(as.data.frame((otu_table(GMD_fec))))
var3 = sample_data(GMD_fec)
treat3 = as.character(var3$ADG_Class)

x3 <- aldex(mi3, treat3, mc.samples=128, test="t", effect=TRUE,
            include.sample.summary=TRUE, denom="iqlr", verbose=TRUE)

aldex.fec=cbind(x3, tax_table(GMD_fec))

res.fec=(aldex.fec[(aldex.fec$effect<="-0.5"),])

Genus = res.fec$Genus
res.fec = cbind(Genus, res.fec)
rownames(res.fec) = NULL
c1 = res.fec[,1]
c2 = res.fec[,3:4]
c3 = res.fec[,33:35]
res3 = cbind(c1,c2,c3)
colnames(res3)
colnames(res3) = c("Genus", "rab.High", "rab.Low", "effect", "overlap", "we.ep") 

round_df <- function(x, digits) {
  # round all numeric variables
  # x: data frame 
  # digits: number of digits to round
  numeric_columns <- sapply(x, mode) == 'numeric'
  x[numeric_columns] <-  round(x[numeric_columns], digits)
  x
}

res4.fec = round_df(res3, 2)
res4.fec




```

```{r}
### Permanovas

## Saliva

set.seed(4040)

saliva = subset_samples(inputR, SampleType == "Saliva")
saliva = subset_samples(saliva, ADG_Class != "NA")
sal_clr <- microbiome::transform(saliva, "clr")

df = as(sample_data(sal_clr), "data.frame")
ds = phyloseq::distance(sal_clr, method = "euclidean")
permanova = adonis(ds ~ ADG_Class, data = df, permutations = 999)

kable(permanova$aov.tab[ ,1:6], caption = "PERMANOVA", format = "markdown")

# Ploting beta diversity graph
input_ord = ordinate(sal_clr, "PCoA" , "euclidean") 
p1 = plot_ordination(sal_clr, input_ord, color = "ADG_Class")
p.sal = p1 + geom_point(size = 6, alpha = 0.5) +
  theme(legend.position = "top") +
  labs(subtitle = "R-squared = 0.06; p = 0.44", color='ADG_Class')+
  labs(title = "Saliva")
p.sal


## Rumen 

rumen = subset_samples(inputR, SampleType == "Ruminal_Fluid")
rumen = subset_samples(rumen, ADG_Class != "NA")
rum_clr <- microbiome::transform(rumen, "clr")

df = as(sample_data(rum_clr), "data.frame")
ds = phyloseq::distance(rum_clr, method = "euclidean")
permanova = adonis(ds ~ ADG_Class, data = df, permutations = 999)

kable(permanova$aov.tab[ ,1:6], caption = "PERMANOVA", format = "markdown")

# Ploting beta diversity graph
input_ord = ordinate(rum_clr, "PCoA" , "euclidean") 
p1 = plot_ordination(rum_clr, input_ord, color = "ADG_Class")
p.rum = p1 + geom_point(size = 6, alpha = 0.5) +
  theme(legend.position = "top") +
  labs(subtitle = "R-squared = 0.13; p = 0.32", color='ADG_Class')+
  labs(title = "Rumen")
p.rum


## Feces

feces = subset_samples(inputR, SampleType == "Feces")
feces = subset_samples(feces, ADG_Class != "NA")
rum_clr <- microbiome::transform(feces, "clr")

df = as(sample_data(rum_clr), "data.frame")
ds = phyloseq::distance(rum_clr, method = "euclidean")
permanova = adonis(ds ~ ADG_Class, data = df, permutations = 999)

kable(permanova$aov.tab[ ,1:6], caption = "PERMANOVA", format = "markdown")

# Ploting beta diversity graph
input_ord = ordinate(sal_clr, "PCoA" , "euclidean") 
p1 = plot_ordination(sal_clr, input_ord, color = "ADG_Class")
p.fec = p1 + geom_point(size = 6, alpha = 0.5) +
  theme(legend.position = "top") +
  labs(subtitle = "R-squared = 0.03; p = 0.06", color='ADG_Class')+
  labs(title = "Feces")
p.fec


p.ast = ggarrange(p.sal, p.rum, p.fec, common.legend = T, ncol = 3)
p.ast

#dev.print(tiff, "C:/Users/Anderson Freitas/Desktop/Seqs/Figures/7_permanova_gmd.tiff",
#          compression = "lzw", res=600, height=8, width=10, units="in")
```


```{r}
input.food <- subset_samples(inputR, Average_CPI != "NA")
input.food <- subset_samples(input.food, ADG_Class != "NA")

cor.eat = sample_data(input.food) %>% as("matrix") %>% data.frame()

ADG = as.numeric(cor.eat$ADG)
Average_CPI = as.numeric(cor.eat$Average_CPI)
ADG_Class = cor.eat$ADG_Class
cor.food = cbind(ADG, Average_CPI, ADG_Class)
cor.food = cor.food %>% as("matrix") %>% data.frame()

cor.food$ADG = as.numeric(cor.food$ADG)
cor.food$Average_CPI = as.numeric(cor.food$Average_CPI)

#View(cor.eat)

plot.food = ggscatter(cor.food, y = "ADG", x = "Average_CPI", 
          add = "reg.line", conf.int = TRUE, size = 4,
          cor.coef = TRUE, cor.method = "spearman",
          xlab = "Crude Protein Intake (g/day)", ylab = "Average Daily Gain of Weight (g/day)",
          add.params = list(color = "blue", fill = "lightgray"),
          color= "ADG_Class")

plot.food

#dev.print(tiff, "E:/Anderson-BackUp/Seqs/Version_Out2021/Figures/Fig4_New.tiff", compression = "lzw", res=600, height=8, width=12, units="in")
```
